tasks/task_manager.cc

```diff
@@ -17,6 +17,62 @@ namespace tasks {
 
 logging::logger tmlogger("task_manager");
 
+bool task_manager::task::children::all_finished() const noexcept {
+    return _children.empty();
+}
+
+size_t task_manager::task::children::size() const noexcept {
+    return _children.size() + _finished_children.size();
+}
+
+future<> task_manager::task::children::add_child(foreign_task_ptr task) {
+    rwlock::holder exclusive_holder = co_await _lock.hold_write_lock();
+
+    auto id = task->id();
+    const auto& [_, inserted] = _children.emplace(id, std::move(task));
+    assert(inserted);
+}
+
+future<> task_manager::task::children::mark_as_finished(task_id id, task_essentials essentials) const {
+    rwlock::holder exclusive_holder = co_await _lock.hold_write_lock();
+
+    auto it = _children.find(id);
+    _finished_children.push_back(essentials);
+    [&] () noexcept {   // erase is expected to not throw
+        _children.erase(it);
+    }();
+}
+
+future<task_manager::task::progress> task_manager::task::children::get_progress(const std::string& progress_units) const {
+    rwlock::holder shared_holder = co_await _lock.hold_read_lock();
+
+    tasks::task_manager::task::progress progress{};
+    co_await coroutine::parallel_for_each(_children, [&] (const auto& child_entry) -> future<> {
+        const auto& child = child_entry.second;
+        auto local_progress = co_await smp::submit_to(child.get_owner_shard(), [&child, &progress_units] {
+            assert(child->get_status().progress_units == progress_units);
```

Comment:
okay, I hope this holds water (and that finished tasks don't change their units when done.
if there's a problem with it in the future, we can go units-less at the aggregation level and have our progress be
`(sum(child->get_progress().completed / child->get_progress().total) + _finished_children.size()) / (_children.size() + _finished_children.size())`